### PR TITLE
Feat/nft whitelist

### PR DESCRIPTION
**Pull request by NelsonGaldeman**

### CODE REVIEW

File: packages/contracts/tests/OrderFactory.t.sol [hunk 1]
```diff
@@ -93,6 +97,47 @@ contract OrderFactoryTest is Test {
     assertEq(sellToken.balanceOf(address(factory)), (_amount * _fee) / HUNDRED_PERCENT);
   }
 
+  function testCreateOrderWithNonceFailWhitelist() public {
+    // burn the whitelist NFT
+    whitelistNFT.burn(1);
+
+    assertEq(factory.whitelist, true);
```
Comment (berteotti) [inline packages/contracts/tests/OrderFactory.t.sol]:
```suggestion
    assertEq(factory.whitelist(), true);
```

Missing parentheses so assert can assess the value.

File: packages/contracts/tests/OrderFactory.t.sol [hunk 2]
```diff
@@ -93,6 +97,47 @@ contract OrderFactoryTest is Test {
...
+  
+  function testCreateOrderWithNonceDisabledWhitelist() public {
+    // Burn whitelist NFT
+    whitelistNFT.burn(1);
+
+    // Disable whitelist
+    factory.toggleWhitelist();
+
+    // Check whitelist was disabled
+    assertEq(factory.whitelist, false);
```
Comment (berteotti) [inline packages/contracts/tests/OrderFactory.t.sol]:
```suggestion
    assertEq(factory.whitelist, false);
```

Here too missing parentheses.

File: packages/contracts/tests/OrderFactory.t.sol [hunk 3]
```diff
@@ -93,6 +97,47 @@ contract OrderFactoryTest is Test {
     assertEq(sellToken.balanceOf(address(factory)), (_amount * _fee) / HUNDRED_PERCENT);
```
Comment (berteotti) [inline packages/contracts/tests/OrderFactory.t.sol:97]: This assertion is failing because _fee isn't equal to contract protocolFee.

**Review by berteotti — CHANGES_REQUESTED** [review]
There are a few tests failing. Otherwise, looks good to me :)

File: packages/contracts/src/OrderFactory.sol [hunk 4]
```diff
@@ -3,19 +3,27 @@ pragma solidity 0.8.20;
...
 error ForbiddenValue();
+error NotWhitelisted();
 
 contract OrderFactory is Ownable2Step {
   using SafeERC20 for IERC20;
 
   uint16 private constant HUNDRED_PERCENT = 10000;
   uint16 public protocolFee = 25; // default 0.25% (range: 0-500 / 0-5%)
+  IERC721 public whitelistNFT;
+  bool public whitelist = true;
```
Comment (berteotti) [inline packages/contracts/src/OrderFactory.sol:19]: Should we call this `openWhitelist`? `whitelist` seems to vague for me

File: packages/contracts/tests/OrderFactory.t.sol (hunk 1 above)
Comment (NelsonGaldeman) [reply packages/contracts/tests/OrderFactory.t.sol]: Fixed thanks!

File: packages/contracts/src/OrderFactory.sol (hunk 4 above)
Comment (NelsonGaldeman) [reply packages/contracts/src/OrderFactory.sol:19]: It feels pretty clear to me tbf. It's a boolean called whitelist so it's `whitelist yes` or `whitelist no`

File: packages/contracts/src/OrderFactory.sol [hunk 5]
```diff
@@ -3,19 +3,27 @@ pragma solidity 0.8.20;
...
 import {SafeERC20} from "oz/token/ERC20/utils/SafeERC20.sol";
 
 error ForbiddenValue();
+error NotWhitelisted();
 
 contract OrderFactory is Ownable2Step {
   using SafeERC20 for IERC20;
 
   uint16 private constant HUNDRED_PERCENT = 10000;
   uint16 public protocolFee = 5; // default 0.05% (range: 0-500 / 0-5%)
```
Comment (berteotti) [inline packages/contracts/src/OrderFactory.sol]: Is this suppose to be 0.05% or 0.25% when deployed?

Comment (NelsonGaldeman) [reply packages/contracts/src/OrderFactory.sol]: 0.25%, but that's not related to this PR

**Review by berteotti — APPROVED** [review]
Good job Neeel!

**Pull request merged**